### PR TITLE
Fixes for lineinfo emission

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Includes: 2d63784 from https://github.com/numba/numba/pull/7177 as 3631bdb

### CODE REVIEW

**Comment by sklam** [comment]
I can confirmed that the decref is no longer jumping back to the line of object creation. I used the following code to test the decref DWARF:

```python


@njit(debug=True)
def foo(n):
    #gdb()
    ar = np.arange(n)
    if ar.sum() == 0:    # lifetime of ar ends after this line
        ret = 2
    else:
        ret = 3
    return ret

```

**Comment by stuartarchibald** [comment]
@sklam thanks for checking. I've patched the disassembly support here: https://github.com/stuartarchibald/numba/commit/ee91dd89747953fe9e0f21c336ed763c5e68e752 to include DWARF info in the disassembly CFG.

**Comment by sklam** [comment]
Cool, i'll try it now

File: numba/tests/test_debuginfo.py [hunk 1]
```diff
@@ -78,6 +78,25 @@ def _get_metadata(self, fn, sig):
...
+        env_copy = os.environ.copy()
+        env_copy['SUBPROC_TEST'] = '1'
+        env_copy['NUMBA_OPT'] = '0'
+        status = subprocess.run(cmd, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE, timeout=60,
+                                env=env_copy, universal_newlines=True)
+        self.assertEqual(status.returncode, 0)
+        self.assertIn('OK', status.stderr)
+        self.assertTrue('FAIL' not in status.stderr)
+        self.assertTrue('ERROR' not in status.stderr)
```
Comment (sklam) [inline numba/tests/test_debuginfo.py:98]: a duplication of https://github.com/numba/numba/blob/0c499bfff7ebe4fe5d8a6c1d20653e69f1f2a639/numba/tests/test_parfors.py#L73-L86

we should dedup here or in a immediate follow up PR

File: numba/tests/support.py [hunk 2]
```diff
@@ -502,6 +510,46 @@ def run_nullary_func(self, pyfunc, flags):
...
+        env_copy.update(envvars)
+        status = subprocess.run(cmd, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE, timeout=timeout,
+                                env=env_copy, universal_newlines=True)
+        streams = (f'\ncaptured stdout: {status.stdout}\n'
+                   f'captured stderr: {status.stderr}')
+        self.assertEqual(status.returncode, 0, streams)
+        self.assertIn('OK', status.stderr)
+        self.assertTrue('FAIL' not in status.stderr)
+        self.assertTrue('ERROR' not in status.stderr)
```
Comment (sklam) [inline numba/tests/support.py:551]:
```suggestion
        self.assertNotIn('ERROR', status.stderr)
```

File: numba/tests/test_debuginfo.py [hunk 3]
```diff
@@ -263,24 +265,121 @@ def test_DILocation_entry_blk(self):
...
+        full_ir = self._get_llvmir(foo, sig=sig)
+
+        # make sure decref lines end with `meminfo.<number>)`.
+        count = 0
+        for line in full_ir.splitlines():
+            line_stripped = line.strip()
+            if line_stripped.startswith('call void @NRT_decref'):
+                self.assertRegex(line, r'.*meminfo\.[0-9]+\)$')
+                count += 1
+        self.assertTrue(count > 0) # make sure there were some decrefs!
```
Comment (sklam) [inline numba/tests/test_debuginfo.py:305]:
```suggestion
        self.assertGreater(count, 0) # make sure there were some decrefs!
```

File: numba/tests/test_debuginfo.py [hunk 4]
```diff
@@ -263,24 +265,121 @@ def test_DILocation_entry_blk(self):
...
+            if a[0] == 0:
+                sink(x)
+                return 12
+            z = x[0][0]
+            return z
+
+        sig = (types.float64[::1],)
+        full_ir = self._get_llvmir(foo, sig=sig)
+
+        # make sure decref lines end with `meminfo.<number>)`.
```
Comment (sklam) [inline numba/tests/test_debuginfo.py:298]:
```suggestion
        # make sure decref lines end with `meminfo.<number>)` without dbg info.
```

**Comment by sklam** [comment]
I wonder if we should leave a note about how the tests are sensitive to SSA and lowering behavior such that they may become outdated when those are changed.

**Comment by stuartarchibald** [comment]
> I wonder if we should leave a note about how the tests are sensitive to SSA and lowering behavior such that they may become outdated when those are changed.

Good point, I've added one as part of 8f1c198. All other comments should be addressed in there too. Thanks for the review.

**Review by sklam — APPROVED** [review]
Thanks for the patch!

**Pull request merged**